apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -0,0 +1,254 @@
+      iex> run_battle(team_1, team_2)
+      :team_1
+  """
+  def run_battle(team_1, team_2, seed \\ 1) do
+    :rand.seed(:default, seed)
+    team_1 = Enum.into(team_1, %{}, fn unit -> create_unit_map(unit, 1) end)
+    team_2 = Enum.into(team_2, %{}, fn unit -> create_unit_map(unit, 2) end)
+    units = Map.merge(team_1, team_2)
+
+    # The initial_step_state is what allows the battle to be simultaneous. If we refreshed the accum on every action,
+    # we would be left with a turn-based battle. Instead we take decisions based on the state of the battle at the beggining
+    # of the step regardless of the changes that happened "before" (execution-wise) in this step.
+    Enum.reduce_while(1..@maximum_steps, units, fn step, initial_step_state ->
+      new_state =
+        initial_step_state
+        |> Enum.map(fn {id, _unit} -> id end)
+        |> Enum.take_random(Enum.count(units))
+        |> Enum.reduce(initial_step_state, fn unit_id, current_state ->
+          process_step(initial_step_state[unit_id], initial_step_state, current_state)
+        end)
+
+      remove_dead_units(new_state)
+      |> check_winner(step)
+    end)
+  end
+
+  defp process_step(unit, initial_step_state, current_state) do
+    new_state =
+      cond do
+        not can_attack(unit) ->
+          current_state
+
+        can_cast_ultimate_skill(unit) ->
+          unit
+          |> cast_skill(unit.ultimate_skill, initial_step_state, current_state)
+          |> put_in([unit.id, :energy], 0)
+
+        can_cast_basic_skill(unit) ->
+          new_state = cast_skill(unit, unit.basic_skill, initial_step_state, current_state)
+
+          new_state
+          |> put_in(
+            [unit.id, :basic_skill, :remaining_cooldown],
+            unit.basic_skill.base_cooldown + 1
+          )
+          |> put_in([unit.id, :energy], new_state[unit.id][:energy] + 75)
+
+        true ->
+          current_state
+      end
+
+    put_in(
+      new_state,
+      [unit.id, :basic_skill, :remaining_cooldown],
+      max(new_state[unit.id].basic_skill.remaining_cooldown - 1, 0)
+    )
+
+    # TODO: decrease_effects_remaining_steps(unit, current_state)
+  end
+
+  defp remove_dead_units(units),
+    do: Enum.filter(units, fn {_id, unit} -> unit.health > 0 end) |> Enum.into(%{})
+
+  defp check_winner(units, step) do
+    winner =
+      cond do
+        Enum.empty?(units) -> :tie
+        Enum.all?(units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
+        Enum.all?(units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
+        true -> :none
+      end
+
+    case winner do
+      :none ->
+        if step == @maximum_steps,
+          do: {:halt, :timeout},
+          else: {:cont, units}
+
+      result ->
+        {:halt, result}
+    end
+  end
+
+  # TODO: Is immobilized?
+  defp can_attack(_unit), do: true
+
+  # Has enough energy?
+  defp can_cast_ultimate_skill(unit), do: unit.energy >= 500
```

Comment:
Maybe this could be stored in a module attribute instead being here

Reply:
Done!